documentation/Investigating-FIR-as-an-MLIR-dialect.md

```diff
@@ -0,0 +1,404 @@
+     any properties of interest. For example, one can define an
+     operation called "execute" and define it to have an attribute
+     "order", which may have a value that describes a sequential
+     execution order, a fully parallel unordered execution, or some
+     valid execution partial ordering. What meta information is
+     interesting depends on the context and is up to the MLIR user.
+
+This document describes FIR, which is a set of extensions upon the
+standard MLIR dialect. When lowering a Fortran parse tree to FIR,
+the compiler will produce a mix of operations defined in the FIR
+dialect, the MLIR __standard dialect__, and other pre-defined
+dialects, such as the [affine dialect](https://bit.ly/2XYDPNj).
+
+#### Disclaimers
+
+This document is not a tutorial on compilers, Fortran, MLIR, LLVM,
+intermediate representations, abstraction levels, semantics, nor type
+theory.
+
+## Requirements Unchanged
+
+The FIR dialect has the same requirements as spelled out in the
+Fortran IR design document. Specifically, the FIR dialect will
+capture the control flow structure of the Fortran source-level
+program. At the highest level of abstraction, Fortran computations
+will be captured as coarse-grain opaque
+operations corresponding to Fortran expressions
+(with links to `Fortran::evaluate::Expr<T>` objects from the
+front-end), where only the peripheral use-def information is
+exposed. These operations will, of course, necessarily be lowered
+to other various MLIR dialects, where they can be
+optimized.
+The compiler will continue to lower the representation
+of the input with successive transformations
+to MLIR's LLVM IR dialect, the target independent LLVM IR, machine
+IR, and eventually assembly and/or machine code.
+
+## Details
+
+Because MLIR is our adopted framework, our bridge to the FIR
+dialect of MLIR will use the concepts, libraries, coding
+conventions, etc. of the MLIR project. This means some of the class
+names will necessarily be changed from the original FIR design. For
+example, `Program` becomes `Module`, `Procedure` becomes
+`Function`, `BasicBlock` becomes `Block`, and `Statement` becomes
+`Operation`.
+
+Construction of the CFG structure reuses the original FIR pass that
+flattens and linearizes the Fortran parse tree structure prior to
+creating the FIR dialect of MLIR.
+
+`Afforestation` (a class in the original FIR code)
+of the FIR dialect tree structure has to be rewritten
+as the original framework is replaced by that of
+MLIR. Conceptually, the process and objective remains the same.
+
+### FIR Dialect Details
+
+The FIR dialect is a well-defined set of operations, types, etc. that
+captures the executable semantics and state of a correctly specified
+Fortran program. In Fortran, action statements [Fortran 2018 R515] and
+certain constructs specify the behavior of the program. Among these,
+Fortran expressions [Fortran 2018 Clause 10.1] are intrinsic to the
+computation of new values.
+
+#### FIR Operations
```

Comment:
Would it be possible to provide a textual representation of the Ops or a an example?